lib/IMAP/MessageMapper.php

```diff
@@ -178,6 +178,20 @@ function (int $uid) use ($highestKnownUid) {
 				return $uid > $highestKnownUid;
 			}
 		);
+		if(count($uidCandidates) === 0) {
```

Comment:
I would like to understand how we can possibly end up here.

1) The mailbox is not empty, or we'll already exit at the `$total === 0` check
2) There was likely an earlier sync chunk that did not reach `$upper === $max`

Does that mean that the *highest known UID* info is possibly wrong when we fetch if from the database? Do you use any kind of fancy database cluster or is it a simple single node database?

Reply:
I reproduced the error by removing my fix and enhancing the log output in the error case (`count($uidCandidates) === 0`) - this is the output: 

`Built range for findAll: min=1 max=35791 total=32122 totalRange=35791 estimatedPageSize=5572 lower=35792 upper=35791 highestKnownUid=35791 count(fetchResult)=1`

And another one:

`Built range for findAll: min=1083 max=1583 total=8 totalRange=501 estimatedPageSize=313126 lower=1584 upper=1583 highestKnownUid=1583 count(fetchResult)=1`

So in both cases the `$upper` is less than `$lower`, while the IMAP query still returns 1 result. I think the IMAP server is smart enough to correct the query by flipping the upper and lower bound. The resulting query can only return the UID equal to highestKnownUid (the other candidate is already > max). The single result then gets filtered out by the code starting in line 168, leaving the `$uidCandidates` empty. 

As an alternative to my solution one could also shortcut the whole thing - in line 145 if `$upper < $lower`.

Reply:
I think I finally understood it. That only took a couple of weeks :speak_no_evil: 

In any case, I've added another unit test for your real world example and adjusted the code for it. I liked your idea with the shortcut as it saves one IMAP roundtrip :v: